### ISSUE TITLE
Reconstruction doesn't match input despite high MS-SSIM 

### ISSUE THREAD

**Issue by tommybotch** [AmigoLab/SynthAnatomy#3]
Hi Dan,

Sorry to bother again. I've successfully trained a network using jukebox perceptual loss that scores high on both MS-SSIM and MSE metrics (see below). The same occurs when training for longer (>100 epochs) and with different loss (MSE).

<img width="1075" alt="Screen Shot 2022-11-03 at 10 10 20 PM" src="https://user-images.githubusercontent.com/42591108/199870371-2fd3a3a2-50cf-4c6d-a5dd-0577f792a127.png">

However, the reconstructed images do not resemble the inputs (see example from validation below). This is true for both train and validation images. I find this odd as the accuracy is reportedly high.

<img width="1078" alt="Screen Shot 2022-11-03 at 10 09 28 PM" src="https://user-images.githubusercontent.com/42591108/199870449-2f3b7997-2d8c-4c24-9099-55b2a14ede65.png">

For reference, I also trained a version of your previous VQVAE network (https://github.com/AmigoLab/nmpevqvae) and managed to get qualitatively similar reconstructions (see below).  

<img width="736" alt="Screen Shot 2022-11-03 at 10 29 17 PM" src="https://user-images.githubusercontent.com/42591108/199872194-ccc35a73-7d24-4f32-8176-0eba0dc7d2b7.png">

Do you have any intuition for why this would occur with SynthAnatomy? Do you have any suggestions on how to solve this issue?

Thanks in advance for your time and help,
Tommy

**Comment by danieltudosiu** [comment]
Hi Tommy,

That is to be expected. The original older VQ-VAE uses a hierarchical approach that we decided not to go forward with due to the complicated and costly nature of sampling from it.

From our other discussion #2 I remember that your latent space was quite small [6,8,6] given the fact that you use the following configuration: 
```
--no_levels=4
--downsample_parameters='((4, 2, 1, 1), (4, 2, 1, 1), (4, 2, 1, 1), (4, 2, 1, 1))'
--upsample_parameters='((4, 2, 1, 0, 1), (4, 2, 1, 0, 1), (4, 2, 1, 0, 1), (4, 2, 1, 0, 1))'
--no_res_layers=3
--no_channels=256
--num_embeddings='(2048,)' 
--embedding_dim='(256,)'
```

I would advise you first to try the following configuration and get back to me with the results and then I can help you further:
```
--no_levels=3
--downsample_parameters='((4, 2, 1, 1), (4, 2, 1, 1), (4, 2, 1, 1))'
--upsample_parameters='((4, 2, 1, 0, 1), (4, 2, 1, 0, 1), (4, 2, 1, 0, 1))'
--no_res_layers=3
--no_channels=256
--num_embeddings='(2048,)' 
--embedding_dim='(256,)'
```

The logic is that you do not need to compress so much due to your lower input dimensionality. You are roughly close to 60% size per axis compared to my input data, thus I would advise you to try the 3 levels of downsampling first.

Also, please show me all the configurations next time as well as all the losses and metrics so I get a proper picture of what's happening.

Alternatively, I am more than happy to have a call with you so I get a better understanding of your requirements and data.

Cheers,

Dan

**Comment by danieltudosiu** [comment]
Hi Tommy,

Also, given your low input dimensionality, the MS-SSIM will have a hard time working due to the smaller kernel size that is being used. As you can see in its name  "Metric-MS-SSIM_3-Reconstruction" 3 stands for the kernel size of the MS-SSIM. 

Such a small kernel size won't be able to properly encompass the spatial structure in your images. In the original MS-SSIM paper, they used a kernel size of 11 but those images were close to 200x200 which is not that common in our field of work to have 200^3 images given the GPU memory we work with.

If you want I can help you modify the code such that you get a better metric for your data. We can change from MS-SSIM to SSIM thus the kernel requirements are not so stringent and we could run a higher kernel size resulting in a more informative metric for your use case.

Cheers,

Dan

**Comment by danieltudosiu** [comment]
Hi Tommy,

A small oversight of our initial implementation is that the output is unbounded, which in turn creates issues with MS-SSIM metric. Assuming you have your data normalized in [0,1] you would want to put your '--output_act' argument to 'SIGMOID' such that you bound your inputs to [0,1] or "TANH" for [-1,1] normalized inputs.

Cheers,

Dab

**Comment by tommybotch** [comment]
Hi Dan,

Thank you for your helpful replies! Sorry for not posting the parameters I used - please see below:

```
    --epochs=10 \
    --learning_rate=0.0001 \
    --log_every=1 \
    --checkpoint_every=1 \
    --eval_every=1 \
    --loss='jukebox_perceptual' \
    --batch_size=8 \
    --adversarial_component=False \
    --normalize=True \
    --eval_batch_size=1 \
    --num_workers=4 \
    --prefetch_factor=4 \
    --starting_epoch=0 \
    --network='baseline_vqvae' \
    --no_levels=4 \
    --downsample_parameters='((4,2,1,1),(4,2,1,1),(4,2,1,1),(4,2,1,1))' \
    --upsample_parameters='((4,2,1,0,1),(4,2,1,0,1),(4,2,1,0,1),(4,2,1,0,1))' \
    --no_res_layers=3 \
    --no_channels=256 \
    --codebook_type='ema' \
    --num_embeddings='(1024,)' \
    --embedding_dim='(32,)' \
    --decay='(0.5,)' \
    --commitment_cost='(0.25,)' \
    --max_decay_epochs=100 
```

This makes sense re: the compression. I've continued training to 500 epochs (previous image was reported at 10 steps, which seemed decent for the previous architecture). It appears that we're beginning to get some properties of the image emerging. My previous cutoff may have been presumptuous.

<img width="1074" alt="Screen Shot 2022-11-04 at 7 24 06 AM" src="https://user-images.githubusercontent.com/42591108/199961452-e8069715-396f-4fa6-81f9-09ab3e37b99d.png">

I agree that we compress to quite a low dimensionality; unfortunately, we'd like to initially compare self-attention and linear attention (Performer, etc) at this level and therefore need dimensionality < 512.

Re: MS-SSIM - the current kernel size used defaults to 3 which, from what you noted, seems too low to get an informative metric. I think, according to the output error when I tried a size of 11, is because images need to be larger than 160 to use MS-SSIM (but could be related to the kernel size). Maybe I try a kernel size between 3 and 11? Despite this, pixel-wise MSE is still high - is this to be expected?

As you noted, I do normalize inputs. I will adjust the output activation accordingly (likely to TanH as our data can span from negative to positive). 

Thank you again for your advice and help. Happy to setup a meeting if that would be easier. 

Best,
Tommy

**Comment by danieltudosiu** [comment]
Hi Tommy,

I would advise against going purely using ' --loss='jukebox_perceptual' ' without any form of adversarial loss. This option is there due to the need of running ablation studies for publications. 

> This makes sense re: the compression. I've continued training for more epochs (previous image was reported at 10 steps, which seemed decent for the previous architecture). It appears that we're beginning to get some properties of the image emerging. My previous cutoff may have been presumptuous.

I would argue that those are not even properties of the image per se. They look more like reconstructions that minimize the chance of getting it wrong across the full dataset than reconstructions that are tied to the actual image. 

> I agree that we compress to quite a low dimensionality; unfortunately, we'd like to initially compare self-attention and linear attention (Performer, etc) at this level and therefore need dimensionality < 512.

I would argue that you do not need this level of compression. On a 32GB card, we are using a Performer network on an even high dimensionality of about ~2k (if I remember correctly) without any problems. Those are the papers where we do so  [[1]](https://link.springer.com/chapter/10.1007/978-3-031-18576-2_2).

> Re: MS-SSIM - the current kernel size used defaults to 3 which, from what you noted, seems too low to get an informative metric. I think, according to the output error when I tried a size of 11, is because images need to be larger than 160 to use MS-SSIM (but could be related to the kernel size). Maybe I try a kernel size between 3 and 11? Despite this, pixel-wise MSE is still high - is this to be expected?

Given the fact that the MS-SSIM kernel is 3, I would not even look at it to be honest. What I would do is first change the MS-SSIM to SSIM by creating a new metric or modifiying the MS-SSIM one [here](https://github.com/AmigoLab/SynthAnatomy/blob/main/src/metrics/vqvae.py#L12) and afterwards try and introduce something like G-SSIM [[2]](https://ieeexplore.ieee.org/document/4107183) instead of classical SSIM. 

> As you noted, I do normalize inputs. I will adjust the output activation accordingly (likely to TanH as our data can span from negative to positive).

Be careful with the current data augmentation strategies that are in the codebase as they were not thoroughly validated on the [-1,1] range since all of us worked in [0,1] . This was also signalled to me by the fact that you have some black stripes around the border of the images due to the augmentations. I would either try to put the background to 0 in the image by doing skull stripping (I think you are already doing it but just covering my bases) and afterwards, I would modify [here](https://github.com/AmigoLab/SynthAnatomy/blob/main/src/utils/vqvae.py#L317) to make sure it is 0 or whatever other padding strategies you want.

Hope this helps! Let me know if you need any more help <3 if you prefer we can always have a chat if it is faster for you ;) 

Cheers,

Dan

**Comment by tommybotch** [comment]
Hi Dan,

Would you recommend using Baur loss as you previously do in your paper? We're avoiding using the GAN component as we're trying to minimize the architecture for our downstream purposes. If you think it's valuable to include this component, then I'm more than happy to include it.
 
> I would advise against going purely using ' --loss='jukebox_perceptual' ' without any form of adversarial loss. This option is there due to the need of running ablation studies for publications.

I take this point (e.g., reconstructed properties are non-image specific - all reconstructed images seem to have the same properties).

> I would argue that those are not even properties of the image per se. They look more like reconstructions that minimize the chance of getting it wrong across the full dataset than reconstructions that are tied to the actual image.

Awesome, will probably start off with SSIM. Thanks for the tip.

> Given the fact that the MS-SSIM kernel is 3, I would not even look at it to be honest. What I would do is first change the MS-SSIM to SSIM by creating a new metric or modifiying the MS-SSIM one [here](https://github.com/AmigoLab/SynthAnatomy/blob/main/src/metrics/vqvae.py#L12) and afterwards try and introduce something like G-SSIM [[2]](https://ieeexplore.ieee.org/document/4107183) instead of classical SSIM.

I can adjust my normalization to [0,1] as it should convey the same information. Thanks for  the advice!

> Be careful with the current data augmentation strategies that are in the codebase as they were not thoroughly validated on the [-1,1] range since all of us worked in [0,1] . This was also signalled to me by the fact that you have some black stripes around the border of the images due to the augmentations. I would either try to put the background to 0 in the image by doing skull stripping (I think you are already doing it but just covering my bases) and afterwards, I would modify [here](https://github.com/AmigoLab/SynthAnatomy/blob/main/src/utils/vqvae.py#L317) to make sure it is 0 or whatever other padding strategies you want.

I appreciate the help and hope to connect in the future. :) 

Best,
Tommy

**Comment by danieltudosiu** [comment]
Hi Tommy,

> Would you recommend using Baur loss as you previously do in your paper? We're avoiding using the GAN component as we're trying to minimize the architecture for our downstream purposes. If you think it's valuable to include this component, then I'm more than happy to include it.

The adversarial loss would not incur such a big memory footprint. From what I remember the bigger footprint is the Perceptual loss component (that uses LPIPs) behind the scenes. So I would strongly encourage you to use adversarial loss. As per the original work that was the basis of ours, they used a Patch-GAN Discriminator as well [[1]](https://arxiv.org/abs/2012.09841).

> Awesome, will probably start off with SSIM. Thanks for the tip.

Another reason that I would advise you to use [0,1] and a Sigmoid at the end is because we do not normalize to 0-1 internally as the MS-SSIM is configured [here](https://github.com/AmigoLab/SynthAnatomy/blob/main/src/metrics/vqvae.py#L17).

> I can adjust my normalization to [0,1] as it should convey the same information. Thanks for the advice!

Please do so as per my point above.

> I appreciate the help and hope to connect in the future. :)

It would be my pleasure to connect,  you can always find me on LinkedIn or Twitter easily from my GitHub landing page.

Cheers,

Dan
